File: fraud_detection.R
# =============================================================================
# Credit Card Fraud Detection Using Machine Learning
# Group 2 - CSC312: Programming for Data Science using R and Python
# =============================================================================

cat(strrep("#", 62), "\n")
cat("#   CREDIT CARD FRAUD DETECTION - GROUP 2 CSC312 PROJECT   #\n")
cat(strrep("#", 62), "\n\n")

# ── Install & load packages ───────────────────────────────────────────────────
user_lib <- file.path(Sys.getenv("USERPROFILE"), "R", "win-library", "4.6")
dir.create(user_lib, recursive = TRUE, showWarnings = FALSE)
.libPaths(c(user_lib, .libPaths()))

needed <- c("ggplot2", "randomForest", "pROC", "ROSE", "dplyr")
missing <- needed[!sapply(needed, requireNamespace, quietly = TRUE)]
if (length(missing) > 0) {
  cat("Installing packages:", paste(missing, collapse = ", "), "\n")
  install.packages(missing, lib = user_lib,
                   repos = "https://cloud.r-project.org",
                   dependencies = FALSE, quiet = TRUE, type = "binary")
}

suppressPackageStartupMessages({
  library(ggplot2)
  library(randomForest)
  library(rpart)
  library(pROC)
  library(ROSE)
  library(dplyr)
  library(gridExtra)
})

# ── Paths ─────────────────────────────────────────────────────────────────────
WORK_DIR  <- "C:/Users/Hillary Chukwuma/OneDrive/Desktop/UPDATE/csc_project"
DATA_PATH <- file.path(WORK_DIR, "creditcard.csv")
PLOT_DIR  <- file.path(WORK_DIR, "plots_R")
MODEL_DIR <- file.path(WORK_DIR, "models_R")
dir.create(PLOT_DIR,  showWarnings = FALSE)
dir.create(MODEL_DIR, showWarnings = FALSE)

set.seed(42)

# =============================================================================
# STEP 1: LOAD DATASET
# =============================================================================
cat(strrep("=", 60), "\n")
cat("  STEP 1: LOADING DATASET\n")
cat(strrep("=", 60), "\n")

df <- read.csv(DATA_PATH)
cat(sprintf("  Dataset shape      : %d rows x %d columns\n", nrow(df), ncol(df)))
cat(sprintf("  Missing values     : %d\n", sum(is.na(df))))

counts <- table(df$Class)
cat("\n  Class distribution:\n")
cat(sprintf("    Legitimate (0) : %s  (%.2f%%)\n",
            format(counts["0"], big.mark = ","), counts["0"] / nrow(df) * 100))
cat(sprintf("    Fraudulent (1) : %s  (%.4f%%)\n",
            format(counts["1"], big.mark = ","), counts["1"] / nrow(df) * 100))

# =============================================================================
# STEP 2: EXPLORATORY DATA ANALYSIS
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 2: EXPLORATORY DATA ANALYSIS\n")
cat(strrep("=", 60), "\n")

df$Class_label <- ifelse(df$Class == 1, "Fraudulent", "Legitimate")

# 2a. Class distribution
p1 <- ggplot(df, aes(x = Class_label, fill = Class_label)) +
  geom_bar(color = "black") +
  scale_fill_manual(values = c("Legitimate" = "#4682B4", "Fraudulent" = "#DC143C")) +
  labs(title = "Class Distribution", x = "Transaction Type", y = "Count") +
  theme_minimal() + theme(legend.position = "none")

png(file.path(PLOT_DIR, "01_class_distribution.png"), width = 700, height = 500)
print(p1); dev.off()
cat("  Saved: 01_class_distribution.png\n")

# 2b. Amount distribution by class
p2 <- ggplot(df, aes(x = Amount, fill = Class_label)) +
  geom_histogram(bins = 60, alpha = 0.7, position = "identity") +
  scale_fill_manual(values = c("Legitimate" = "#4682B4", "Fraudulent" = "#DC143C")) +
  coord_cartesian(xlim = c(0, 2500)) +
  labs(title = "Transaction Amount Distribution",
       x = "Amount (USD)", y = "Count", fill = "Type") +
  theme_minimal()

png(file.path(PLOT_DIR, "02_amount_distribution.png"), width = 900, height = 500)
print(p2); dev.off()
cat("  Saved: 02_amount_distribution.png\n")

# 2c. Fraud rate by hour
df$Hour <- floor(df$Time / 3600) %% 24
hourly  <- df %>% group_by(Hour) %>%
  summarise(fraud_rate = mean(Class) * 100, .groups = "drop")

p3 <- ggplot(hourly, aes(x = Hour, y = fraud_rate)) +
  geom_line(color = "#DC143C", linewidth = 1) +
  geom_point(color = "#DC143C", size = 2) +
  labs(title = "Fraud Rate by Hour of Day", x = "Hour", y = "Fraud Rate (%)") +
  theme_minimal()

png(file.path(PLOT_DIR, "03_fraud_by_hour.png"), width = 900, height = 450)
print(p3); dev.off()
cat("  Saved: 03_fraud_by_hour.png\n")

# 2d. Top feature correlations with Class
num_df   <- df %>% select(-Class_label, -Hour)
cors     <- cor(num_df)[, "Class"]
cors     <- sort(abs(cors), decreasing = TRUE)
top10    <- names(cors)[2:11]

cat("\n  Top correlated features with fraud:\n")
for (f in top10) cat(sprintf("    %-6s : %.4f\n", f, cors[f]))

cor_df <- data.frame(Feature = top10, Correlation = cors[top10])
p4 <- ggplot(cor_df, aes(x = reorder(Feature, Correlation), y = Correlation)) +
  geom_bar(stat = "identity", fill = "steelblue", color = "black") +
  coord_flip() +
  labs(title = "Top 10 Features Correlated with Fraud",
       x = "Feature", y = "|Correlation|") +
  theme_minimal()

png(file.path(PLOT_DIR, "04_feature_correlation.png"), width = 800, height = 500)
print(p4); dev.off()
cat("  Saved: 04_feature_correlation.png\n")

# =============================================================================
# STEP 3: PREPROCESSING & CLASS IMBALANCE
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 3: PREPROCESSING & HANDLING CLASS IMBALANCE\n")
cat(strrep("=", 60), "\n")

df_clean        <- df %>% select(-Class_label, -Hour)
df_clean$Amount <- as.numeric(scale(df_clean$Amount))
df_clean$Time   <- as.numeric(scale(df_clean$Time))
df_clean$Class  <- as.factor(df_clean$Class)

# Stratified 80/20 split
fraud_idx  <- which(df_clean$Class == 1)
legit_idx  <- which(df_clean$Class == 0)
train_fraud <- sample(fraud_idx, floor(0.8 * length(fraud_idx)))
train_legit <- sample(legit_idx, floor(0.8 * length(legit_idx)))
train_idx   <- c(train_fraud, train_legit)

train_df <- df_clean[train_idx, ]
test_df  <- df_clean[-train_idx, ]
cat(sprintf("  Train size : %d  |  Test size : %d\n", nrow(train_df), nrow(test_df)))

# ROSE oversampling on training set
rose_out <- ROSE(Class ~ ., data = train_df, seed = 42)$data
rose_out$Class <- as.factor(rose_out$Class)
cat(sprintf("  After ROSE – Legit: %d  |  Fraud: %d\n",
            sum(rose_out$Class == 0), sum(rose_out$Class == 1)))

X_test <- test_df %>% select(-Class)
y_test <- as.integer(as.character(test_df$Class))

# =============================================================================
# STEP 4: MODEL TRAINING & EVALUATION
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 4: MODEL TRAINING & EVALUATION\n")
cat(strrep("=", 60), "\n")

# Helper: compute metrics
get_metrics <- function(y_true, y_pred, y_prob) {
  tp <- sum(y_pred == 1 & y_true == 1)
  fp <- sum(y_pred == 1 & y_true == 0)
  fn <- sum(y_pred == 0 & y_true == 1)
  precision <- if ((tp + fp) == 0) 0 else tp / (tp + fp)
  recall    <- if ((tp + fn) == 0) 0 else tp / (tp + fn)
  f1        <- if ((precision + recall) == 0) 0 else
                  2 * precision * recall / (precision + recall)
  roc_obj   <- roc(y_true, y_prob, quiet = TRUE)
  auc_val   <- as.numeric(auc(roc_obj))
  list(precision = precision, recall = recall,
       f1 = f1, auc = auc_val, roc = roc_obj)
}

print_metrics <- function(name, m, y_true, y_pred) {
  cat(sprintf("\n  [%s]\n",   name))
  cat(sprintf("    Precision : %.4f\n", m$precision))
  cat(sprintf("    Recall    : %.4f\n", m$recall))
  cat(sprintf("    F1-Score  : %.4f\n", m$f1))
  cat(sprintf("    ROC-AUC   : %.4f\n", m$auc))
  cat("    Confusion Matrix (rows=Actual, cols=Predicted):\n")
  print(table(Actual = y_true, Predicted = y_pred))
}

# ── Logistic Regression ───────────────────────────────────────────────────────
cat("\n  Training Logistic Regression...\n")
lr_model  <- glm(Class ~ ., data = rose_out, family = binomial)
lr_prob   <- predict(lr_model, X_test, type = "response")
lr_pred   <- as.integer(lr_prob >= 0.5)
lr_m      <- get_metrics(y_test, lr_pred, lr_prob)
print_metrics("Logistic Regression", lr_m, y_test, lr_pred)

# ── Decision Tree ─────────────────────────────────────────────────────────────
cat("\n  Training Decision Tree...\n")
dt_model  <- rpart(Class ~ ., data = rose_out, method = "class",
                   control = rpart.control(maxdepth = 8))
dt_prob   <- predict(dt_model, X_test, type = "prob")[, 2]
dt_pred   <- as.integer(dt_prob >= 0.5)
dt_m      <- get_metrics(y_test, dt_pred, dt_prob)
print_metrics("Decision Tree", dt_m, y_test, dt_pred)

# ── Random Forest ─────────────────────────────────────────────────────────────
cat("\n  Training Random Forest...\n")
rf_model  <- randomForest(Class ~ ., data = rose_out, ntree = 100,
                          importance = TRUE)
rf_prob   <- predict(rf_model, X_test, type = "prob")[, 2]
rf_pred   <- as.integer(rf_prob >= 0.5)
rf_m      <- get_metrics(y_test, rf_pred, rf_prob)
print_metrics("Random Forest", rf_m, y_test, rf_pred)

# =============================================================================
# STEP 5: VISUALISATIONS
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 5: VISUALISATIONS\n")
cat(strrep("=", 60), "\n")

model_names <- c("Logistic Regression", "Decision Tree", "Random Forest")
metrics_all <- list(lr_m, dt_m, rf_m)
colors_vec  <- c("#4C72B0", "#DD8452", "#55A868")

# 5a. Model comparison bar chart
comp_df <- data.frame(
  Model  = rep(model_names, each = 4),
  Metric = rep(c("Precision","Recall","F1","ROC-AUC"), 3),
  Score  = c(
    lr_m$precision, lr_m$recall, lr_m$f1, lr_m$auc,
    dt_m$precision, dt_m$recall, dt_m$f1, dt_m$auc,
    rf_m$precision, rf_m$recall, rf_m$f1, rf_m$auc
  )
)
comp_df$Model  <- factor(comp_df$Model,  levels = model_names)
comp_df$Metric <- factor(comp_df$Metric, levels = c("Precision","Recall","F1","ROC-AUC"))

p5 <- ggplot(comp_df, aes(x = Metric, y = Score, fill = Model)) +
  geom_bar(stat = "identity", position = "dodge", color = "black", alpha = 0.85) +
  scale_fill_manual(values = setNames(colors_vec, model_names)) +
  ylim(0, 1.1) +
  labs(title = "Model Performance Comparison", y = "Score") +
  theme_minimal()

png(file.path(PLOT_DIR, "05_model_comparison.png"), width = 1000, height = 500)
print(p5); dev.off()
cat("  Saved: 05_model_comparison.png\n")

# 5b. ROC Curves
png(file.path(PLOT_DIR, "06_roc_curves.png"), width = 800, height = 600)
plot(lr_m$roc, col = colors_vec[1], lwd = 2, main = "ROC Curves – All Models")
plot(dt_m$roc, col = colors_vec[2], lwd = 2, add = TRUE)
plot(rf_m$roc, col = colors_vec[3], lwd = 2, add = TRUE)
legend("bottomright",
       legend = sprintf("%s (AUC=%.4f)", model_names,
                        sapply(metrics_all, `[[`, "auc")),
       col = colors_vec, lwd = 2, bty = "n")
dev.off()
cat("  Saved: 06_roc_curves.png\n")

# 5c. Confusion matrices
make_cm_plot <- function(y_true, y_pred, title) {
  cm    <- as.data.frame(table(Actual = factor(y_true, c(0,1)),
                               Predicted = factor(y_pred, c(0,1))))
  cm$Actual    <- ifelse(cm$Actual == "1",    "Fraud", "Legit")
  cm$Predicted <- ifelse(cm$Predicted == "1", "Fraud", "Legit")
  ggplot(cm, aes(x = Predicted, y = Actual, fill = Freq)) +
    geom_tile(color = "white") +
    geom_text(aes(label = Freq), size = 6, fontface = "bold") +
    scale_fill_gradient(low = "white", high = "steelblue") +
    labs(title = title) +
    theme_minimal() + theme(legend.position = "none")
}

cm1 <- make_cm_plot(y_test, lr_pred, "Logistic Regression")
cm2 <- make_cm_plot(y_test, dt_pred, "Decision Tree")
cm3 <- make_cm_plot(y_test, rf_pred, "Random Forest")

png(file.path(PLOT_DIR, "07_confusion_matrices.png"), width = 1200, height = 400)
gridExtra::grid.arrange(cm1, cm2, cm3, ncol = 3)
dev.off()
cat("  Saved: 07_confusion_matrices.png\n")

# 5d. Random Forest feature importance
imp_df <- data.frame(
  Feature    = rownames(importance(rf_model)),
  Importance = importance(rf_model)[, "MeanDecreaseGini"]
)
imp_df <- imp_df[order(imp_df$Importance, decreasing = TRUE), ][1:15, ]

p6 <- ggplot(imp_df, aes(x = reorder(Feature, Importance), y = Importance)) +
  geom_bar(stat = "identity", fill = "steelblue", color = "black") +
  coord_flip() +
  labs(title = "Top 15 Feature Importances – Random Forest",
       x = "Feature", y = "Mean Decrease Gini") +
  theme_minimal()

png(file.path(PLOT_DIR, "08_feature_importance.png"), width = 900, height = 500)
print(p6); dev.off()
cat("  Saved: 08_feature_importance.png\n")

# =============================================================================
# STEP 6: SELECT BEST MODEL & SAVE
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 6: MODEL SELECTION & SAVING\n")
cat(strrep("=", 60), "\n")

f1_scores <- c(lr_m$f1, dt_m$f1, rf_m$f1)
best_idx  <- which.max(f1_scores)
best_name <- model_names[best_idx]
best_model <- list(lr_model, dt_model, rf_model)[[best_idx]]
best_m    <- metrics_all[[best_idx]]

cat(sprintf("\n  Best model (by F1-Score) : %s\n", best_name))
cat(sprintf("    F1      : %.4f\n", best_m$f1))
cat(sprintf("    ROC-AUC : %.4f\n", best_m$auc))

model_path <- file.path(MODEL_DIR, "best_model.rds")
saveRDS(best_model, model_path)
cat(sprintf("  Saved to : %s\n", model_path))

# =============================================================================
# STEP 7: PREDICTION SYSTEM DEMO
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  STEP 7: PREDICTION SYSTEM DEMO\n")
cat(strrep("=", 60), "\n")

predict_transaction <- function(model, txn_row, model_type = "rf") {
  if (model_type == "lr") {
    prob <- predict(model, txn_row, type = "response")[[1]]
  } else if (model_type == "dt") {
    prob <- predict(model, txn_row, type = "prob")[1, 2]
  } else {
    prob <- predict(model, txn_row, type = "prob")[1, 2]
  }
  label <- ifelse(prob >= 0.5, "FRAUDULENT", "LEGITIMATE")
  list(label = label, probability = prob)
}

fraud_rows <- which(y_test == 1)
legit_rows <- which(y_test == 0)

for (info in list(
  list(idx = fraud_rows[1], true = "FRAUDULENT", type = "rf"),
  list(idx = legit_rows[1], true = "LEGITIMATE",  type = "rf")
)) {
  txn    <- X_test[info$idx, , drop = FALSE]
  result <- predict_transaction(best_model, txn, info$type)
  match  <- ifelse(result$label == info$true, "CORRECT", "WRONG")
  cat(sprintf("\n  True label : %s\n",         info$true))
  cat(sprintf("  Prediction : %s  (prob=%.4f)  [%s]\n",
              result$label, result$probability, match))
}

# =============================================================================
# FINAL SUMMARY
# =============================================================================
cat("\n", strrep("=", 60), "\n")
cat("  FINAL SUMMARY\n")
cat(strrep("=", 60), "\n")

summary_df <- data.frame(
  Model     = model_names,
  Precision = round(sapply(metrics_all, `[[`, "precision"), 4),
  Recall    = round(sapply(metrics_all, `[[`, "recall"),    4),
  F1        = round(sapply(metrics_all, `[[`, "f1"),        4),
  ROC_AUC   = round(sapply(metrics_all, `[[`, "auc"),       4)
)
print(summary_df, row.names = FALSE)
cat(sprintf("\n  Best Model  : %s\n", best_name))
cat(sprintf("  Plots saved : %s\n",  PLOT_DIR))
cat(sprintf("  Model saved : %s\n",  model_path))
cat(strrep("=", 60), "\n\n")

File: fraud_detection.py
"""
Credit Card Fraud Detection Using Machine Learning
Group 2 - CSC Project
"""

import warnings
warnings.filterwarnings("ignore")

import os
import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

from sklearn.model_selection import train_test_split, StratifiedKFold, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import (
    classification_report, confusion_matrix, roc_auc_score,
    roc_curve, precision_recall_curve, f1_score, precision_score, recall_score
)
from imblearn.over_sampling import SMOTE
from imblearn.under_sampling import RandomUnderSampler

# ─── Paths ────────────────────────────────────────────────────────────────────
BASE_DIR  = os.path.dirname(os.path.abspath(__file__))
DATA_PATH = os.path.join(BASE_DIR, "creditcard.csv")
PLOT_DIR  = os.path.join(BASE_DIR, "plots")
MODEL_DIR = os.path.join(BASE_DIR, "models")
os.makedirs(PLOT_DIR,  exist_ok=True)
os.makedirs(MODEL_DIR, exist_ok=True)

# ─── 1. Load Data ─────────────────────────────────────────────────────────────
def load_data():
    print("\n" + "="*60)
    print("  STEP 1: LOADING DATASET")
    print("="*60)
    df = pd.read_csv(DATA_PATH)
    print(f"Dataset shape      : {df.shape}")
    print(f"Features           : {list(df.columns)}")
    print(f"Missing values     : {df.isnull().sum().sum()}")
    print(f"\nClass distribution :")
    counts = df["Class"].value_counts()
    print(f"  Legitimate (0)   : {counts[0]:,}  ({counts[0]/len(df)*100:.2f}%)")
    print(f"  Fraudulent (1)   : {counts[1]:,}  ({counts[1]/len(df)*100:.4f}%)")
    return df

# ─── 2. EDA ───────────────────────────────────────────────────────────────────
def exploratory_data_analysis(df):
    print("\n" + "="*60)
    print("  STEP 2: EXPLORATORY DATA ANALYSIS")
    print("="*60)

    # 2a. Class distribution bar chart
    fig, axes = plt.subplots(1, 2, figsize=(12, 4))
    counts = df["Class"].value_counts()
    axes[0].bar(["Legitimate", "Fraudulent"], counts.values,
                color=["steelblue", "crimson"], edgecolor="black")
    axes[0].set_title("Class Distribution")
    axes[0].set_ylabel("Number of Transactions")
    for i, v in enumerate(counts.values):
        axes[0].text(i, v + 500, f"{v:,}", ha="center", fontweight="bold")

    # 2b. Transaction amount distribution
    axes[1].hist(df[df["Class"]==0]["Amount"], bins=60, alpha=0.6,
                 label="Legitimate", color="steelblue")
    axes[1].hist(df[df["Class"]==1]["Amount"], bins=60, alpha=0.8,
                 label="Fraudulent", color="crimson")
    axes[1].set_title("Transaction Amount Distribution")
    axes[1].set_xlabel("Amount (USD)")
    axes[1].set_ylabel("Count")
    axes[1].legend()
    axes[1].set_xlim(0, 2500)
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "01_class_and_amount.png"), dpi=150)
    plt.close()
    print("  Saved: 01_class_and_amount.png")

    # 2c. Correlation heatmap (top 15 features by absolute correlation with Class)
    corr = df.corr()["Class"].drop("Class").abs().sort_values(ascending=False)
    top_feats = corr.head(15).index.tolist()
    fig, ax = plt.subplots(figsize=(10, 8))
    sns.heatmap(df[top_feats + ["Class"]].corr(), annot=True, fmt=".2f",
                cmap="coolwarm", ax=ax, linewidths=0.5)
    ax.set_title("Correlation Heatmap – Top 15 Features vs Class")
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "02_correlation_heatmap.png"), dpi=150)
    plt.close()
    print("  Saved: 02_correlation_heatmap.png")

    # 2d. Time vs fraud rate
    df_copy = df.copy()
    df_copy["Hour"] = (df_copy["Time"] // 3600) % 24
    hourly = df_copy.groupby("Hour")["Class"].mean() * 100
    fig, ax = plt.subplots(figsize=(10, 4))
    ax.plot(hourly.index, hourly.values, marker="o", color="crimson")
    ax.set_title("Fraud Rate by Hour of Day")
    ax.set_xlabel("Hour")
    ax.set_ylabel("Fraud Rate (%)")
    ax.grid(True, alpha=0.3)
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "03_fraud_by_hour.png"), dpi=150)
    plt.close()
    print("  Saved: 03_fraud_by_hour.png")

    print(f"\n  Top correlated features with fraud:")
    for feat, val in corr.head(10).items():
        print(f"    {feat:<6}: {val:.4f}")

# ─── 3. Preprocessing ─────────────────────────────────────────────────────────
def preprocess(df):
    print("\n" + "="*60)
    print("  STEP 3: PREPROCESSING & HANDLING CLASS IMBALANCE")
    print("="*60)

    # Scale Amount and Time
    scaler = StandardScaler()
    df = df.copy()
    df["Amount"] = scaler.fit_transform(df[["Amount"]])
    df["Time"]   = scaler.fit_transform(df[["Time"]])

    X = df.drop("Class", axis=1)
    y = df["Class"]

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42, stratify=y
    )
    print(f"  Train size: {X_train.shape[0]:,}  |  Test size: {X_test.shape[0]:,}")

    # SMOTE on training set only
    smote = SMOTE(random_state=42)
    X_res, y_res = smote.fit_resample(X_train, y_train)
    print(f"  After SMOTE  – Legit: {(y_res==0).sum():,}  |  Fraud: {(y_res==1).sum():,}")

    # Also create undersampled set for comparison
    rus = RandomUnderSampler(random_state=42)
    X_under, y_under = rus.fit_resample(X_train, y_train)
    print(f"  After UnderSampling – Legit: {(y_under==0).sum():,}  |  Fraud: {(y_under==1).sum():,}")

    return X_train, X_test, y_train, y_test, X_res, y_res, X_under, y_under, X.columns.tolist()

# ─── 4. Train & Evaluate Models ───────────────────────────────────────────────
def train_and_evaluate(name, model, X_train, y_train, X_test, y_test):
    model.fit(X_train, y_train)
    y_pred  = model.predict(X_test)
    y_proba = model.predict_proba(X_test)[:, 1]

    precision = precision_score(y_test, y_pred)
    recall    = recall_score(y_test, y_pred)
    f1        = f1_score(y_test, y_pred)
    roc_auc   = roc_auc_score(y_test, y_proba)

    print(f"\n  [{name}]")
    print(f"    Precision : {precision:.4f}")
    print(f"    Recall    : {recall:.4f}")
    print(f"    F1-Score  : {f1:.4f}")
    print(f"    ROC-AUC   : {roc_auc:.4f}")
    print(classification_report(y_test, y_pred,
                                target_names=["Legitimate", "Fraudulent"],
                                digits=4))
    return model, y_pred, y_proba, {
        "Model": name, "Precision": precision,
        "Recall": recall, "F1": f1, "ROC_AUC": roc_auc
    }

def run_models(X_res, y_res, X_test, y_test, feature_names):
    print("\n" + "="*60)
    print("  STEP 4: MODEL TRAINING & EVALUATION (SMOTE data)")
    print("="*60)

    models_def = {
        "Logistic Regression": LogisticRegression(max_iter=1000, random_state=42),
        "Decision Tree":       DecisionTreeClassifier(max_depth=8, random_state=42),
        "Random Forest":       RandomForestClassifier(n_estimators=100, random_state=42, n_jobs=-1),
    }

    results  = []
    trained  = {}
    preds    = {}

    for name, mdl in models_def.items():
        mdl, y_pred, y_proba, metrics = train_and_evaluate(
            name, mdl, X_res, y_res, X_test, y_test
        )
        trained[name] = mdl
        preds[name]   = (y_pred, y_proba)
        results.append(metrics)

    return trained, preds, results

# ─── 5. Visualise Results ─────────────────────────────────────────────────────
def visualise_results(trained, preds, results, X_test, y_test, feature_names):
    print("\n" + "="*60)
    print("  STEP 5: VISUALISATIONS")
    print("="*60)

    colors = ["#4C72B0", "#DD8452", "#55A868"]

    # 5a. Model comparison bar chart
    df_res = pd.DataFrame(results)
    metrics = ["Precision", "Recall", "F1", "ROC_AUC"]
    x = np.arange(len(metrics))
    width = 0.25
    fig, ax = plt.subplots(figsize=(11, 5))
    for i, (_, row) in enumerate(df_res.iterrows()):
        ax.bar(x + i * width, [row[m] for m in metrics],
               width, label=row["Model"], color=colors[i], alpha=0.85, edgecolor="black")
    ax.set_xticks(x + width)
    ax.set_xticklabels(metrics)
    ax.set_ylim(0, 1.1)
    ax.set_ylabel("Score")
    ax.set_title("Model Performance Comparison")
    ax.legend()
    ax.grid(axis="y", alpha=0.3)
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "04_model_comparison.png"), dpi=150)
    plt.close()
    print("  Saved: 04_model_comparison.png")

    # 5b. ROC curves
    fig, ax = plt.subplots(figsize=(8, 6))
    for (name, (_, y_proba)), color in zip(preds.items(), colors):
        fpr, tpr, _ = roc_curve(y_test, y_proba)
        auc = roc_auc_score(y_test, y_proba)
        ax.plot(fpr, tpr, label=f"{name} (AUC={auc:.4f})", color=color, lw=2)
    ax.plot([0,1],[0,1], "k--", lw=1)
    ax.set_xlabel("False Positive Rate")
    ax.set_ylabel("True Positive Rate")
    ax.set_title("ROC Curves – All Models")
    ax.legend()
    ax.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "05_roc_curves.png"), dpi=150)
    plt.close()
    print("  Saved: 05_roc_curves.png")

    # 5c. Precision-Recall curves
    fig, ax = plt.subplots(figsize=(8, 6))
    for (name, (_, y_proba)), color in zip(preds.items(), colors):
        prec, rec, _ = precision_recall_curve(y_test, y_proba)
        ax.plot(rec, prec, label=name, color=color, lw=2)
    ax.set_xlabel("Recall")
    ax.set_ylabel("Precision")
    ax.set_title("Precision-Recall Curves – All Models")
    ax.legend()
    ax.grid(alpha=0.3)
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "06_precision_recall_curves.png"), dpi=150)
    plt.close()
    print("  Saved: 06_precision_recall_curves.png")

    # 5d. Confusion matrices
    fig, axes = plt.subplots(1, 3, figsize=(15, 4))
    for ax, (name, (y_pred, _)) in zip(axes, preds.items()):
        cm = confusion_matrix(y_test, y_pred)
        sns.heatmap(cm, annot=True, fmt="d", cmap="Blues", ax=ax,
                    xticklabels=["Legit", "Fraud"],
                    yticklabels=["Legit", "Fraud"])
        ax.set_title(name)
        ax.set_ylabel("Actual")
        ax.set_xlabel("Predicted")
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "07_confusion_matrices.png"), dpi=150)
    plt.close()
    print("  Saved: 07_confusion_matrices.png")

    # 5e. Random Forest feature importance
    rf = trained["Random Forest"]
    importances = pd.Series(rf.feature_importances_, index=feature_names)
    top15 = importances.sort_values(ascending=False).head(15)
    fig, ax = plt.subplots(figsize=(9, 5))
    top15.sort_values().plot.barh(ax=ax, color="steelblue", edgecolor="black")
    ax.set_title("Top 15 Feature Importances – Random Forest")
    ax.set_xlabel("Importance Score")
    plt.tight_layout()
    plt.savefig(os.path.join(PLOT_DIR, "08_feature_importance.png"), dpi=150)
    plt.close()
    print("  Saved: 08_feature_importance.png")

    return df_res

# ─── 6. Select Best Model & Save ──────────────────────────────────────────────
def select_and_save(trained, df_res):
    print("\n" + "="*60)
    print("  STEP 6: MODEL SELECTION & SAVING")
    print("="*60)
    best_row   = df_res.loc[df_res["F1"].idxmax()]
    best_name  = best_row["Model"]
    best_model = trained[best_name]

    print(f"\n  Best model (by F1-Score): {best_name}")
    print(f"    F1      : {best_row['F1']:.4f}")
    print(f"    ROC-AUC : {best_row['ROC_AUC']:.4f}")

    path = os.path.join(MODEL_DIR, "best_model.pkl")
    joblib.dump(best_model, path)
    print(f"  Saved to : {path}")
    return best_name, best_model

# ─── 7. Prediction System ─────────────────────────────────────────────────────
def predict_transaction(model, transaction_dict, feature_names):
    """
    Accepts a dict of feature values and returns fraud probability + label.
    """
    row = pd.DataFrame([transaction_dict], columns=feature_names)
    prob  = model.predict_proba(row)[0][1]
    label = "FRAUDULENT" if prob >= 0.5 else "LEGITIMATE"
    return label, prob

def demo_prediction(best_model, X_test, y_test, feature_names):
    print("\n" + "="*60)
    print("  STEP 7: PREDICTION SYSTEM DEMO")
    print("="*60)

    # Pick a known fraud and a known legitimate transaction
    fraud_idx  = y_test[y_test == 1].index[0]
    legit_idx  = y_test[y_test == 0].index[0]

    for idx, true_label in [(fraud_idx, "FRAUDULENT"), (legit_idx, "LEGITIMATE")]:
        txn = X_test.loc[idx].to_dict()
        pred_label, prob = predict_transaction(best_model, txn, feature_names)
        match = "CORRECT" if pred_label == true_label else "WRONG"
        print(f"\n  True label : {true_label}")
        print(f"  Prediction : {pred_label}  (probability={prob:.4f})  [{match}]")

# ─── 8. Summary ───────────────────────────────────────────────────────────────
def print_summary(df_res, best_name):
    print("\n" + "="*60)
    print("  FINAL SUMMARY")
    print("="*60)
    print(df_res.to_string(index=False))
    print(f"\n  Best Model : {best_name}")
    print(f"  All plots saved to  : {PLOT_DIR}")
    print(f"  Best model saved to : {MODEL_DIR}/best_model.pkl")
    print("="*60 + "\n")

# ─── Main ─────────────────────────────────────────────────────────────────────
def main():
    print("\n" + "#"*60)
    print("#   CREDIT CARD FRAUD DETECTION – GROUP 2 CSC PROJECT    #")
    print("#"*60)

    df = load_data()
    exploratory_data_analysis(df)
    X_train, X_test, y_train, y_test, X_res, y_res, X_under, y_under, feat_names = preprocess(df)
    trained, preds, results = run_models(X_res, y_res, X_test, y_test, feat_names)
    df_res = visualise_results(trained, preds, results, X_test, y_test, feat_names)
    best_name, best_model = select_and_save(trained, df_res)
    demo_prediction(best_model, X_test, y_test, feat_names)
    print_summary(df_res, best_name)

if __name__ == "__main__":
    main()
